### ISSUE TITLE
Unnecessary training on the final token in SFT

### ISSUE THREAD

**Issue by ad8e** [THUDM/slime#863]
https://github.com/THUDM/slime/blob/70d65e06cf9fcfd8f9064a885ec88227b38f9949/slime/backends/megatron_utils/loss.py#L65

Existing behavior: the model receives all tokens, and the last token is sliced out after the forward is run, in the above line.

The correct way:

`tokens`, a tensor of length 100, is received from the dataset
`inputs`, a tensor of length 99, consisting of `tokens[:-1]`, is received by the model
`logits`, a tensor of length 99, is produced by the model.

Then the model handles 100-1 tokens, saving 1 token of context length.

**Comment by rbao2018** [comment]
The final eos token must be input into the label. Removing the last eos token from the input will prevent the model from learning the eos token. As a result, the model will be unable to generate the eos token to stop during deployment.

**Comment by ad8e** [comment]
The eos token only needs to be in the label, not the input. Training does not need to predict the token after the eos token.
